src/main/java/com/hierynomus/msdfsc/ReferralCache.java

```diff
@@ -183,4 +170,46 @@ public String toString() {
 
     }
 
+
+    static class ReferralCacheNode {
+        String pathComponent;
+        Map<String, ReferralCacheNode> childNodes = new HashMap<>();
+        ReferralCacheEntry entry;
+
+        ReferralCacheNode(String pathComponent) {
+            this.pathComponent = pathComponent;
+        }
+
+        ReferralCacheNode(String pathComponent, ReferralCacheEntry entry) {
+            this.pathComponent = pathComponent;
+            this.entry = entry;
+        }
+
+        void addReferralEntry(Iterator<String> pathComponents, ReferralCacheEntry entry) {
+            if (pathComponents.hasNext()) {
+                String component = pathComponents.next();
+                if (!childNodes.containsKey(component)) {
```

Comment:
```
ReferralCacheNode node = childNodes.get(component);
if (node == null) {
    node = new ReferralCacheNode(component);
    childNodes.put(component, node);
}
node.addReferralEntry(pathComponents, entry);
```

That'll be slightly faster. `contains` followed by `get` means you effectively do the same hash table lookup twice.